ADR/nnnn-trusted-task-model.md

```diff
@@ -0,0 +1,238 @@
+- Denial takes priority
+  - Even if a Task is on the `trusted_tasks` list, a deny rule should be able to
+    make it untrusted.
+  - Even if an allow rule makes a Task trusted, if it's expired according to the
+    `trusted_tasks` list, it should be considered untrusted.
+
+This ADR doesn't mandate a specific format for `trusted_task_rules`, but does
+come with an example of what it could look like - see the [Appendix](#appendix).
+
+### Trust upstream Konflux Tasks based on bundle URL
+
+A policy author who wants to trust all the upstream Konflux Tasks should do so by
+defining a rule that allows all task bundles from `quay.io/konflux-ci/tekton-catalog/*`
+with no further conditions. The `trusted_task_rules` mechanism must support this.
+
+### Start signing upstream Konflux tasks with a common release key
+
+Trusting all task bundles from `quay.io/konflux-ci/tekton-catalog/*` is not a strong
+trust model (but isn't weaker than the current one - see [Consequences](#consequences)).
+To let policy authors define a stronger trust model, we should aim to start signing
+the upstream Konflux task bundles with a common release key and extend the
+`trusted_task_rules` mechanism to support signature verification.
+
+## Consequences
+
+### `data-acceptable-bundles` no longer blocks decentralization
+
+Decentralized Konflux Tasks are trusted implicitly, as long as they release to the
+right location.
+
+### Strength of trust model for upstream Konflux Tasks
+
+Current trust model, in theory:
+
+- We place trust in the `quay.io/konflux-ci/tekton-catalog/data-acceptable-bundles`
+  artifact
+  - Only those who can write to this artifact can make Tasks trusted
+
+Current trust model, in practice:
+
+- Who can write to `quay.io/konflux-ci/tekton-catalog/data-acceptable-bundles`:
+  - Release pipelines in the Red Hat deployment of Konflux
+  - The CI pipelines for build-definitions
+  - The CI pipelines for other repos that happen to use the same OpenShift namespace
+    as build-definitions
+
+The proposed trust model, without signing key verification:
+
+- We implicitly trust all tasks in `quay.io/konflux-ci/tekton-catalog/*`
+- We trust anyone who can write to any of those repositories
+  - Which is the same list as for `data-acceptable-bundles`
+
+With signing key verification:
+
+- We trust only the release pipelines in the Red Hat deployment of Konflux
+
+### Git references to upstream Konflux Tasks no longer trusted
+
+In the current model, any change that gets merged to build-definitions is considered
+trusted, so the `trusted_tasks` data includes both `git` references and `oci` bundle
+references.
+
+The new trust model for upstream Konflux Tasks is based around trusting the bundles
+released through Red Hat's deployment of Konflux. We should not try to establish
+trust for the `git` references to the source code for those tasks, because there will
+be no way to know whether the reference passed release-time verifications or not.
```

Comment:
I'd love to keep the option to still trust git references. Can you think of a reasonable way to still make it possible?

What if the release process signed git tags and pushed them back to the git repository?

We could leave this out of scope for _this_ decision and follow up with another one later that adds this capability.

I know, at least in the Fedora space, users have expressed a desire to see git references rather than bundle references because they're easier to follow as a user to understand what's happening. Easier to hack on.

Reply:
One of the sticky points with git is that for any single file, it could have the same exact content across different commit IDs. (A new commit ID can modify other files, for example.) In this repo, we chose to only use the [git commit ID that modified the file](https://github.com/konflux-ci/build-definitions/blob/8fcc48dc4f857db446be0c92c544937fc36ad6e0/.tekton/scripts/build-acceptable-bundles.sh#L26). It's not very intuitive but it mostly works :)

With the caveat above, using signatures on git commits is probably a good way forward to establish trust, but this would have to be done by an external entity, like a release pipeline. It's definitely worth investigating more.

Reply:
Rephrased some parts of this section and mentioned the suggested approaches (signed tags, separate "release repo") as possible future directions